File: worker/src/routes/unsubscribe.ts
import { Hono } from 'hono';
import { Env } from '../env';
import { InvalidTokenError, verify } from '../lib/token';
import { UnsubscribePage } from '../pages/unsubscribe';
import { getContactByID } from '../store/contacts';
import { getListByID } from '../store/lists';
import { getSend } from '../store/sends';
import {
  getSubscriptionByID,
  unsubscribeSubscription,
} from '../store/subscriptions';
import { NotFoundError } from '../store/types';
import { recordUnsubscribeEvent } from '../store/unsubs';

function htmlResponse(html: any, status = 200) {
  return new Response(String(html), {
    status,
    headers: { 'content-type': 'text/html; charset=utf-8' },
  });
}

async function verifyTurnstile(env: Env, captchaToken: string, ip: string): Promise<boolean> {
  if (!env.MINIGUN_TURNSTILE_SECRET_KEY) return true;
  const form = new FormData();
  form.append('secret', env.MINIGUN_TURNSTILE_SECRET_KEY);
  form.append('response', captchaToken);
  if (ip) form.append('remoteip', ip);
  const resp = await fetch('https://challenges.cloudflare.com/turnstile/v0/siteverify', {
    method: 'POST',
    body: form,
  });
  if (!resp.ok) return false;
  const data = (await resp.json()) as { success?: boolean };
  return data.success === true;
}

function clientIP(headers: Headers): string {
  const xff = headers.get('x-forwarded-for');
  if (xff) return xff.split(',')[0]!.trim();
  return headers.get('cf-connecting-ip') ?? headers.get('x-real-ip') ?? '';
}

export function mountUnsubscribe(app: Hono<{ Bindings: Env }>) {
  app.get('/u/:token', async (c) => {
    const tokenStr = c.req.param('token');
    try {
      const t = await verify(c.env.MINIGUN_HMAC_SECRET, tokenStr);
      const snd = await getSend(c.env.DB, t.sendID);
      const sub = await getSubscriptionByID(c.env.DB, t.subscriptionID);
      const contact = await getContactByID(c.env.DB, sub.contact_id);
      let listName = '';
      if (snd.list_id) {
        try {
          listName = (await getListByID(c.env.DB, snd.list_id)).name;
        } catch {}
      }
      if (!sub.subscribed) {
        return htmlResponse(
          UnsubscribePage({ email: contact.email, listName, token: tokenStr, done: true }),
        );
      }
      return htmlResponse(
        UnsubscribePage({
          email: contact.email,
          listName,
          token: tokenStr,
          turnstileSiteKey: c.env.MINIGUN_TURNSTILE_SITE_KEY,
        }),
      );
    } catch (err) {
      const msg =
        err instanceof InvalidTokenError
          ? 'Invalid or expired unsubscribe link.'
          : err instanceof NotFoundError
            ? 'Subscription not found.'
            : 'Invalid or expired unsubscribe link.';
      return htmlResponse(
        UnsubscribePage({ email: '', listName: '', token: tokenStr, error: msg }),
        400,
      );
    }
  });

  app.post('/u/:token', async (c) => {
    const tokenStr = c.req.param('token');
    const form = (await c.req.parseBody().catch(() => ({}))) as Record<string, string | File>;
    const oneClick =
      typeof form['List-Unsubscribe'] === 'string' && form['List-Unsubscribe'] === 'One-Click';

    let t;
    try {
      t = await verify(c.env.MINIGUN_HMAC_SECRET, tokenStr);
    } catch {
      if (oneClick) return new Response(null, { status: 400 });
      return htmlResponse(
        UnsubscribePage({
          email: '',
          listName: '',
          token: tokenStr,
          error: 'Invalid or expired unsubscribe link.',
        }),
        400,
      );
    }

    if (!oneClick && c.env.MINIGUN_TURNSTILE_SITE_KEY) {
      const captcha = typeof form['cf-turnstile-response'] === 'string' ? form['cf-turnstile-response'] : '';
      const ok = await verifyTurnstile(c.env, captcha, clientIP(c.req.raw.headers));
      if (!ok) {
        return htmlResponse(
          UnsubscribePage({
            email: '',
            listName: '',
            token: tokenStr,
            error: 'Bot challenge failed. Please try again.',
            turnstileSiteKey: c.env.MINIGUN_TURNSTILE_SITE_KEY,
          }),
          400,
        );
      }
    }

    let snd;
    try {
      snd = await getSend(c.env.DB, t.sendID);
    } catch {
      if (oneClick) return new Response(null, { status: 200 });
      return htmlResponse(
        UnsubscribePage({
          email: '',
          listName: '',
          token: tokenStr,
          error: 'Send not found.',
        }),
        400,
      );
    }

    let sub;
    try {
      sub = await getSubscriptionByID(c.env.DB, t.subscriptionID);
    } catch {
      if (oneClick) return new Response(null, { status: 200 });
      return htmlResponse(
        UnsubscribePage({
          email: '',
          listName: '',
          token: tokenStr,
          error: 'Already unsubscribed.',
        }),
        400,
      );
    }
    if (sub.subscribed) {
      sub = await unsubscribeSubscription(c.env.DB, sub.list_id, sub.contact_id);
    }
    let email = '';
    try {
      email = (await getContactByID(c.env.DB, sub.contact_id)).email;
    } catch {}
    if (email) {
      await recordUnsubscribeEvent(c.env.DB, snd.id, sub, email).catch((err) =>
        console.error('record unsub event', err),
      );
    }

    if (oneClick) return new Response(null, { status: 200 });

    if (snd.unsubscribe_mode === 'external' && snd.unsubscribe_external_url) {
      return Response.redirect(withUnsubQuery(snd.unsubscribe_external_url, sub.id, snd.list_id), 302);
    }
    if ((snd.unsubscribe_mode as string) === 'redirect' && snd.unsubscribe_redirect_url) {
      return Response.redirect(withUnsubQuery(snd.unsubscribe_redirect_url, sub.id, snd.list_id), 302);
    }

    let listName = '';
    if (snd.list_id) {
      try {
        listName = (await getListByID(c.env.DB, snd.list_id)).name;
      } catch {}
    }
    return htmlResponse(UnsubscribePage({ email, listName, token: tokenStr, done: true }));
  });
}

function withUnsubQuery(base: string, subID: number, listID?: string | null): string {
  try {
    const u = new URL(base);
    if (listID) u.searchParams.set('list', listID);
    u.searchParams.set('subscription_id', String(subID));
    return u.toString();
  } catch {
    return base;
  }
}

File: src/internal/api/unsubscribe.go
package api

import (
	"context"
	"errors"
	"net/http"
	"net/url"
	"strings"

	"github.com/go-chi/chi/v5"

	"github.com/ranaroussi/minigun/internal/models"
	"github.com/ranaroussi/minigun/internal/store"
	"github.com/ranaroussi/minigun/internal/tmpl"
	"github.com/ranaroussi/minigun/internal/token"
)

func (s *Server) handleUnsubscribeGet(w http.ResponseWriter, r *http.Request) {
	tokenStr := chi.URLParam(r, "token")
	t, err := token.Verify(s.cfg.HMACSecret, tokenStr)
	if err != nil {
		s.renderUnsubPage(w, tmpl.UnsubscribeData{Error: "Invalid or expired unsubscribe link."})
		return
	}
	snd, err := s.store.GetSend(r.Context(), t.SendID)
	if err != nil {
		if errors.Is(err, store.ErrNotFound) {
			s.renderUnsubPage(w, tmpl.UnsubscribeData{Done: true})
			return
		}
		s.renderUnsubPageStatus(w, http.StatusInternalServerError, tmpl.UnsubscribeData{Error: "Something went wrong. Please try again."})
		return
	}
	sub, err := s.store.GetSubscriptionByID(r.Context(), t.SubscriptionID)
	if err != nil {
		if errors.Is(err, store.ErrNotFound) {
			s.renderUnsubPage(w, tmpl.UnsubscribeData{Done: true})
			return
		}
		s.renderUnsubPageStatus(w, http.StatusInternalServerError, tmpl.UnsubscribeData{Error: "Something went wrong. Please try again."})
		return
	}
	contact, err := s.store.GetContactByID(r.Context(), sub.ContactID)
	if err != nil {
		if errors.Is(err, store.ErrNotFound) {
			s.renderUnsubPage(w, tmpl.UnsubscribeData{Done: true})
			return
		}
		s.renderUnsubPageStatus(w, http.StatusInternalServerError, tmpl.UnsubscribeData{Error: "Something went wrong. Please try again."})
		return
	}
	listName := ""
	if snd.ListID != nil {
		if l, err := s.store.GetListByID(r.Context(), *snd.ListID); err == nil {
			listName = l.Name
		}
	}

	if !sub.Subscribed {
		s.renderUnsubPage(w, tmpl.UnsubscribeData{
			Done:     true,
			Email:    contact.Email,
			ListName: listName,
		})
		return
	}

	s.renderUnsubPage(w, tmpl.UnsubscribeData{
		Token:            tokenStr,
		Email:            contact.Email,
		ListName:         listName,
		TurnstileSiteKey: s.cfg.TurnstileSiteKey,
	})
}

func (s *Server) handleUnsubscribePost(w http.ResponseWriter, r *http.Request) {
	tokenStr := chi.URLParam(r, "token")
	if err := r.ParseForm(); err != nil {
		writeError(w, http.StatusBadRequest, err.Error())
		return
	}

	oneClick := isRFC8058OneClick(r)

	t, err := token.Verify(s.cfg.HMACSecret, tokenStr)
	if err != nil {
		if oneClick {
			w.WriteHeader(http.StatusBadRequest)
			return
		}
		s.renderUnsubPage(w, tmpl.UnsubscribeData{Error: "Invalid or expired unsubscribe link."})
		return
	}

	if !oneClick && s.cfg.TurnstileSiteKey != "" {
		captchaToken := r.FormValue("cf-turnstile-response")
		ip := clientIP(r)
		res, err := s.turnstile.Verify(r.Context(), captchaToken, ip)
		if err != nil || res == nil || !res.Success {
			s.renderUnsubPage(w, tmpl.UnsubscribeData{
				Token: tokenStr,
				Error: "Bot challenge failed. Please try again.",
				TurnstileSiteKey: s.cfg.TurnstileSiteKey,
			})
			return
		}
	}

	snd, err := s.store.GetSend(r.Context(), t.SendID)
	if err != nil {
		if oneClick {
			w.WriteHeader(http.StatusOK)
			return
		}
		if errors.Is(err, store.ErrNotFound) {
			s.renderUnsubPage(w, tmpl.UnsubscribeData{Done: true})
			return
		}
		s.renderUnsubPageStatus(w, http.StatusInternalServerError, tmpl.UnsubscribeData{Error: "Something went wrong. Please try again."})
		return
	}

	sub, err := s.unsubAndRecord(r.Context(), t, snd)
	if err != nil {
		if oneClick {
			w.WriteHeader(http.StatusOK)
			return
		}
		s.renderUnsubPage(w, tmpl.UnsubscribeData{Done: true})
		return
	}

	if oneClick {
		w.WriteHeader(http.StatusOK)
		return
	}

	switch snd.UnsubscribeMode {
	case models.UnsubModeRedirect:
		if snd.UnsubscribeRedirectURL != nil {
			http.Redirect(w, r, withUnsubQuery(*snd.UnsubscribeRedirectURL, sub, snd), http.StatusFound)
			return
		}
	case models.UnsubModeExternal:
		if snd.UnsubscribeExternalURL != nil {
			http.Redirect(w, r, withUnsubQuery(*snd.UnsubscribeExternalURL, sub, snd), http.StatusFound)
			return
		}
	}

	listName := ""
	email := ""
	if c, err := s.store.GetContactByID(r.Context(), sub.ContactID); err == nil {
		email = c.Email
	}
	if snd.ListID != nil {
		if l, err := s.store.GetListByID(r.Context(), *snd.ListID); err == nil {
			listName = l.Name
		}
	}
	s.renderUnsubPage(w, tmpl.UnsubscribeData{
		Done:     true,
		Email:    email,
		ListName: listName,
	})
}

func (s *Server) unsubAndRecord(ctx context.Context, t *token.Unsubscribe, snd *models.Send) (*models.Subscription, error) {
	sub, err := s.store.GetSubscriptionByID(ctx, t.SubscriptionID)
	if err != nil {
		return nil, err
	}
	if sub.Subscribed {
		updated, err := s.store.UnsubscribeSubscription(ctx, sub.ListID, sub.ContactID)
		if err != nil {
			return nil, err
		}
		sub = updated
	}
	contact, err := s.store.GetContactByID(ctx, sub.ContactID)
	if err == nil {
		_, _ = s.store.RecordUnsubscribeEvent(ctx, &snd.ID, sub, contact.Email)
	}
	return sub, nil
}

func (s *Server) renderUnsubPage(w http.ResponseWriter, data tmpl.UnsubscribeData) {
	w.Header().Set("Content-Type", "text/html; charset=utf-8")
	if data.Error != "" {
		w.WriteHeader(http.StatusBadRequest)
	}
	if err := tmpl.Unsubscribe.Execute(w, data); err != nil {
		s.log.Error("render unsub page", "err", err)
	}
}

func (s *Server) renderUnsubPageStatus(w http.ResponseWriter, status int, data tmpl.UnsubscribeData) {
	w.Header().Set("Content-Type", "text/html; charset=utf-8")
	w.WriteHeader(status)
	if err := tmpl.Unsubscribe.Execute(w, data); err != nil {
		s.log.Error("render unsub page", "err", err)
	}
}

func isRFC8058OneClick(r *http.Request) bool {
	if v := r.FormValue("List-Unsubscribe"); v == "One-Click" {
		return true
	}
	return false
}

func clientIP(r *http.Request) string {
	if ip := r.Header.Get("X-Forwarded-For"); ip != "" {
		parts := strings.Split(ip, ",")
		return strings.TrimSpace(parts[0])
	}
	if ip := r.Header.Get("X-Real-IP"); ip != "" {
		return ip
	}
	addr := r.RemoteAddr
	if i := strings.LastIndex(addr, ":"); i > 0 {
		return addr[:i]
	}
	return addr
}

func withUnsubQuery(base string, sub *models.Subscription, snd *models.Send) string {
	u, err := url.Parse(base)
	if err != nil {
		return base
	}
	q := u.Query()
	if snd.ListID != nil {
		q.Set("list", *snd.ListID)
	}
	q.Set("subscription_id", itoa(sub.ID))
	u.RawQuery = q.Encode()
	return u.String()
}

func itoa(n int64) string {
	if n == 0 {
		return "0"
	}
	neg := false
	if n < 0 {
		neg = true
		n = -n
	}
	var buf [20]byte
	i := len(buf)
	for n > 0 {
		i--
		buf[i] = byte('0' + n%10)
		n /= 10
	}
	if neg {
		i--
		buf[i] = '-'
	}
	return string(buf[i:])
}

var _ = errors.New
